Support SK2 in Basic Configure Functions

fire-at-will opened the pull request:
### Motivation
One of the main goals of the 5.0.0 release of `purchases-ios` is to handle purchases with StoreKit 2 by default. This PR gets us closer to that goal by introducing new parameters to the simpler SDK `configure` functions that support StoreKit 2 by default when possible.

### Changes Introduced
- Introduced a new parameter, `storeKitVersion: StoreKitVersion` to the non-builder observer mode configure function.
   - A default value is not provided, forcing developers to choose their StoreKit version.

- Obsoleted the old observer mode configure function without the `storeKitVersion` parameter.
   - Developers using `Purchases.configure(withAPIKey:appUserID:observerMode:)` will hit the old obsoleted configure function, and will be pushed to specify a StoreKit version:
<img width="1185" alt="Screenshot 2024-04-19 at 3 29 18 PM" src="https://github.com/RevenueCat/purchases-ios/assets/7330140/e2842c30-17c9-47e0-aaf4-5c88d01f2cf9">

- Updated the Swift & Objective-C API testers to account for the new changes

- Updated to our DocC documentation to reflect these new function signatures

- Included a warning in `StoreKitVersion.storeKit2`'s documentation instructing developers using observer mode to call `Purchases/handleObserverModeTransaction(_:)` after making a purchase.

### Breaking Changes
- Developers that were using `Purchases.configure(withAPIKey:appUserID:observerMode:)` will be forced to migrate to `Purchases.configure(withAPIKey:appUserID:observerMode:storeKitVersion)`.

### Other Approaches Considered
See [this Google doc](https://docs.google.com/document/d/1Kwri7qcSMQRlyNyn4Dyckqxd8eVf_JqEkH1fJJsk994/edit#heading=h.m1h0u03xo2zt) for a list of other approaches considered.

aboedo commented on Sources/DocCDocumentation/DocCDocumentation.docc/Purchases.md:
```diff
@@ -22,7 +22,7 @@ Most features require configuring the SDK before using it.
 
 ### Configuring the SDK
-- ``Purchases/configure(withAPIKey:)``
+- ``Purchases/configure(withAPIKey:storeKitVersion:)``
```
honestly I think this kind of defeats the purpose of having these functions in the first place - they're meant to keep configuration as easy as possible for the majority of the customer base. 

Most customers should not have to think about the storekit version, they pay us to do it for them.

aboedo commented on Sources/Misc/Obsoletions.swift:
```diff
@@ -526,6 +526,115 @@ public extension Purchases {
+     */
+    @objc(configureWithAPIKey:)
+    @available(iOS, obsoleted: 1, message: """
+    If you're using observer mode with StoreKit 2, ensure that you call Purchases.handleObserverModeTransaction(result:) after making a purchase.
```
this note is great, let's also add it to the docstring though. In the docstring we can even use double backticks syntax to create a link to the other method

aboedo commented on Sources/Misc/Obsoletions.swift:
```diff
@@ -526,6 +526,115 @@ public extension Purchases {
+     *
+     * - Returns: An instantiated ``Purchases`` object that has been set as a singleton.
+     */
+    @objc(configureWithAPIKey:)
```
I don't understand why we need to break this API - this is for folks who don't use observer mode. Why should they have to change anything? Isn't it fine for them to just switch seamlessly from SK1 to SK2?

aboedo commented on Sources/Purchasing/Purchases/Purchases.swift:
```diff
@@ -1312,20 +1322,22 @@ public extension Purchases {
      * RevenueCat's backend. Default is `false`.
      *
+     * - Parameter storeKitVersion: The StoreKit version Purchases will use to process your purchases.
+     *
      * - Returns: An instantiated ``Purchases`` object that has been set as a singleton.
      *
      * - Warning: This assumes your IAP implementation uses StoreKit 1.
      */
-    @_disfavoredOverload
-    @objc(configureWithAPIKey:appUserID:observerMode:)
+    @objc(configureWithAPIKey:appUserID:observerMode:storeKitVersion:)
     @discardableResult static func configure(withAPIKey apiKey: String,
                                              appUserID: String?,
-                                             observerMode: Bool) -> Purchases {
+                                             observerMode: Bool,
+                                             storeKitVersion: StoreKitVersion) -> Purchases {
```
this is the only one that I actually understand why we need to break

aboedo commented:
> A default value is not provided for the configure functions that include the ability to set observerMode, forcing developers to choose their StoreKit version.

This seems fine and is consistent with the builder version of it, but we're also not providing a default value for the ones that do not provide the ability to set observer mode, which has me confused

fire-at-will commented:
@aboedo Thanks for the review, your comments make a lot of sense! I've made a few changes based on the feedback:
- Removed the optional parameter `storeKitVersion` from `Purchases.configure(withAPIKey:)` and `Purchases.configure(withAPIKey:appUserID)` and set it so that they use StoreKit 2
   - Updated the docs to reflect these changes
- Remove the obsoletions on the non observer-mode configure functions
- Added the warning on the obsoleted observer mode configure function to its docstring
- Fixed the docstring reference on the observer-mode constructor. Previously, it stated that you could use that function to set a custom UserDefaults, but the function didn't allow developers pass one in.

Now the only breaking change in the PR is for the observer mode configure function. 😎

aboedo commented on Sources/Misc/Obsoletions.swift:
```diff
@@ -526,6 +526,48 @@ public extension Purchases {
+     *
+     * - Returns: An instantiated ``Purchases`` object that has been set as a singleton.
+     *
+     * - Warning: This assumes your IAP implementation uses StoreKit 1.
+     * - Warning: If you're using observer mode with StoreKit 2, ensure that you call ``Purchases/handleObserverModeTransaction(_:)`` after making a purchase.
```
The second warning is kind of a no-op since this API is obsoleted, right? Meaning you can't use it in any case, so you would still have to move to the SK2 version of it? 

Assuming that's correct, maybe we can tell folks to move to that method as well, like
```suggestion
     * - Warning: This assumes your IAP implementation uses StoreKit 1.
     * - Warning: If you're using observer mode with StoreKit 2, configure the SDK with `configure(withAPIKey:appUserID:observerMode:storeKitVersion:)` passing in `.storeKit2` as the `storeKitVersion` and ensure that you call ``Purchases/handleObserverModeTransaction(_:)`` after making a purchase.
```

aboedo commented:
left a minor suggestion but good to go!
